feat: allow custom storage dir overrides

bwl opened the pull request:
The strategy is --storage-dir flag, then TASKBOOK_DIR, then config, then default locations.

This pr reorganizes things just a bit and adds support for the flag and ENV overrides.

I apologize for the architectural rework required to support this feature. The changes convert taskbook from a singleton export to a class that can be instantiated with options. While this is a breaking change at the module level, it does not affect end-users or CLI behavior—existing workflows continue to work exactly as before.

I really love your taskbook ux and it's really nice to use per project, especially with coding agents.

I'm using this alias and it rocks, but i didn't want to propose this change in the core because it work break tb for existing users if they happened to end up in a dir with a .taskbook folder

      ## Shell Alias for Smooth Workflow
      ```bash
      tb() {
        if [ -d "./.taskbook" ]; then
          command tb --storage-dir ./.taskbook "$@"
        else
          command tb "$@"
        fi
      }
      ```

bwl committed:
feat: allow custom storage dir overrides

Honor --storage-dir flag above TASKBOOK_DIR, config, and default locations.

Note: consider adding a shell alias wrapper that checks for a local .taskbook directory and calls 'tb --storage-dir ./.taskbook ' when present for a smoother workflow.

klaudiosinani commented on src/help.js:
```diff
@@ -21,6 +21,7 @@ module.exports = `
       --priority, -p     Update priority of task
       --restore, -r      Restore items from archive
       --star, -s         Star/unstar item
+      --storage-dir      Use custom storage directory
```
A better and more explicit choice could be:

```
--taskbook-directory   Define a custom taskbook directory 
```

klaudiosinani commented on src/storage.js:
```diff
@@ -20,20 +21,63 @@ class Storage {
+    const isTaskbookDir = basename(dir) === '.taskbook';
+    const baseDir = isTaskbookDir ? dirname(dir) : dir;
+    const mainDir = isTaskbookDir ? dir : join(dir, '.taskbook');
+    return {baseDir, mainDir};
```
I think the directory related responsibility is gettting to much for the storage class to handle, a better approach could be to introduce a `directory.js` module, for example:

```js
class Directory {
  _taskbookDirectoryName = '.taskbook';
  _userHomeDirectory = os.homedir();

  get taskbookDirectoryName() {
    return this._taskbookDirectoryName;
  }

  get userConfigTaskbookParentDirectory() {
    const {taskbookDirectory} = config.get();
    return taskbookDirectory;
  }

  get environmentVariableTaskbookParentDirectory() {
    return process.env.TASKBOOK_DIRECTORY;
  }

  retrieveTaskbookDirectory(options) {
    const parent = this._retrieveTaskbookParentDirectory(options);
    return this._composeTaskbookDirectory(parent);
  }

  _retrieveTaskbookParentDirectory(options) {
    const customTaskbookParentDirectory = this._retrieveTaskbookCustomParentDirectory(options);

    if (customTaskbookParentDirectory) {
      return customTaskbookParentDirectory;
    }

    return this._retrieveDefaultTaskbookParentDirectory();
  }

  _retrieveTaskbookCustomParentDirectory(options) {
    const candidates = this._retrieveTaskbookParentDirectoryCandidates(options);
    const presentCandidates = this._filterPresentTaskbookParentDirectoryCandidates(candidates);
    return this._retrieveTaskbookCustomParentDirectoryOrExit(presentCandidates);
  }

  _retrieveTaskbookCustomParentDirectoryOrExit(candidates) {
    const selectedCandidate = this._selectHighestPriorityTaskbookParentDirectory(candidates);
    const isValidCandidate = this._isValidTaskbookCustomParentDirectory(selectedCandidate);

    if (isValidCandidate) {
      return this._parseDirectory(selectedCandidate);
    }

    render.invalidCustomAppDir(selectedCandidate);
    process.exit(1);
  }

  _retrieveTaskbookParentDirectoryCandidates(options) {
    return [
      this._getTaskbookParentDirectoryFlagParameter(options),
      this.environmentVariableTaskbookParentDirectory,
      this.userConfigTaskbookParentDirectory,
    ];
  }

  _getTaskbookParentDirectoryFlagParameter(options) {
    return options.taskbookDirectory;
  }

  _filterPresentTaskbookParentDirectoryCandidates(candidates) {
    return candidates.filter(candidate => this._isDefined(candidate) && this._isStringType(candidate));
  }

  _selectHighestPriorityTaskbookParentDirectory(candidates) {
    return candidates[0];
  }

  _retrieveDefaultTaskbookParentDirectory() {
    return this._userHomeDirectory;
  }

  _composeTaskbookDirectory(parentDirectory) {
    return join(parentDirectory, this.taskbookDirectoryName);
  }

  _isValidTaskbookCustomParentDirectory(directory) {
    return !this._isEmptyString(directory) && this._isExistingDirectory(directory);
  }

  _isExistingDirectory(directory) {
    const parsedDirectory = this._parseDirectory(directory);
    return fs.existsSync(parsedDirectory);
  }

  _isValidDirectoryFormat(directory) {
    return this._isStringType(directory);
  }

  _parseDirectory(directory) {
    const expandedDirectory = this._expandDirectory(directory);
    return resolve(expandedDirectory);
  }

  _isDefined(input) {
    return input !== undefined && input !== null;
  }

  _isStringType(input) {
    return typeof input === 'string';
  }

  _isEmptyString(directory) {
    return this._trim(directory).length === 0;
  }

  _trim(directory) {
    return directory.trim();
  }

  _expandDirectory(directory) {
    return directory.replace(/^~(?=$|[\\/])/, os.homedir());
  }
}
```

The key resposibilities would be:

- enforcing the resolution by priority of the possible custom taskbook parent directories:
  1. CLI: --taskbook-directory
  2. Env Var: process.env.TASKBOOK_DIRECTORY
  3. User config: ~/.taskbook.json > taskbookDirectory
- Ensuring the fallback to the hardcoded user home directory as the parent
- Ensuring the custom directory set by the user is a non empty string that represents a path in their file system; only when one of the 2 predicates is not standing taskbook exits the process

klaudiosinani reviewed (changes requested):
It's a great idea and well done for the first iteration of your implementation

Please have a look at the comments and happy to hear your thoughts

bwl committed:
refactor: extract directory logic per PR #214 review

Address review feedback from @klaudiosinani:

1. Rename CLI flag from --storage-dir to --taskbook-dir
   - More explicit and better aligned in help text
   - Update all references (cli.js, index.js, directory.js)

2. Extract directory resolution into new src/directory.js module
   - Implement Directory class per maintainer's specification
   - Handle priority-based resolution: flag → env var → config → default
   - Validate directories with proper error handling
   - Support tilde expansion and path normalization
   - Storage class now focused solely on file operations

Changes maintain backward compatibility while improving
separation of concerns and code organization.

bwl committed:
fix: use node: prefix for built-in modules

Address SonarCloud code quality issues:
- Use node:fs instead of fs
- Use node:os instead of os
- Use node:path instead of path

Improves code clarity by distinguishing built-in modules
from npm packages per modern Node.js best practices.

klaudiosinani commented on src/directory.js:
```diff
@@ -0,0 +1,132 @@
+  }
+}
+
+module.exports = Directory;
```
Given `directory` is immutable and effectively stateless you could directly export a reference to an instance here and import it in `storage` thus treating it as a singleton

klaudiosinani commented on src/help.js:
```diff
@@ -21,6 +21,7 @@ module.exports = `
       --priority, -p     Update priority of task
       --restore, -r      Restore items from archive
       --star, -s         Star/unstar item
+      --taskbook-dir     Define a custom taskbook directory
```
You are right; given the established standard set by `git --git-dir`, `npm --cache-dir` etc; this is more suitable option of the two

klaudiosinani commented on index.js:
```diff
@@ -1,8 +1,10 @@
+const Taskbook = require('./src/taskbook');
 
 const taskbookCLI = (input, flags) => {
+  const taskbook = new Taskbook({taskbookDir: flags.taskbookDir});
```
why not just pass the full object, so that in the future if we parse more flag data we do not touch this line; what do you think?

bwl replied:
great suggestions, i made the tweaks and it does feel a lot cleaner. Let me know if there is anything else to address here. thanks again.

bwl committed:
refactor: apply singleton pattern and pass full flags object

Address maintainer feedback:

1. Export directory as singleton (src/directory.js)
   - Directory is immutable/stateless, singleton is appropriate
   - Reduces instantiation overhead

2. Import singleton directly in storage (src/storage.js)
   - Remove 'new Directory()' instantiation
   - Use imported singleton reference

3. Pass full flags object to Taskbook (index.js)
   - Change from {taskbookDir: flags.taskbookDir} to flags
   - Future-proof for additional flag parameters

klaudiosinani commented on src/directory.js:
```diff
@@ -0,0 +1,132 @@
+  _retrieveTaskbookCustomParentDirectoryOrExit(candidates) {
+    const selectedCandidate = this._selectHighestPriorityTaskbookParentDirectory(candidates);
+
+    if (!this._isDefined(selectedCandidate)) {
```
this condidational blocks looks like it's dead code given that we already filter for defined candidates on line 44; what do you think?

bwl replied:
great catch. Definitely dead code, I went through and made sure everything is serving a purpose, there were a couple other spots like this.

klaudiosinani commented on src/directory.js:
```diff
@@ -0,0 +1,132 @@
+  }
+
+  get environmentVariableTaskbookParentDirectory() {
+    return process.env.TASKBOOK_DIRECTORY;
```
For consistency we could use `process.env.TASKBOOK_DIR` instead of `process.env.TASKBOOK_DIRECTORY`

bwl replied:
Made this tweak while doing the other work, sorry forgot to commit it separately.

bwl committed:
refactor: improve code formatting and simplify directory validation logic

Removing unused methods _isValidDirectoryFormat, _isEmptyString, and
_trim.
Inline logic of_isValidTaskbookCustomParentDirectoryhas.
Simplify redundant checks in_retrieveTaskbookCustomParentDirectoryOrExit
and _filterPresentTaskbookParentDirectoryCandidates.
Fix minor formatting (spaces inside const { }) issue.

bwl committed:
Update readme to include runtime directory override feature

bwl commented:
Code comments are addressed and a draft of readme changes are up. Would be happy to include more examples and bash snippets that could help people ergonomically use their own custom taskbook directories but didn't want to go too far without your thoughts.

klaudiosinani commented on src/directory.js:
```diff
@@ -0,0 +1,114 @@
+  _retrieveTaskbookCustomParentDirectoryOrExit(candidates) {
+    const selectedCandidate = this._selectHighestPriorityTaskbookParentDirectory(candidates);
+
+    if (!this._isExistingDirectory(selectedCandidate)) {
```
If we check only for existance then the following case will pass as valid instead of rendering an error and exiting:

```
$ tb --taskbook-dir --task hello world
# The effective directory that will be provided will be an empty string {taskbokDir: ""};
# which in turn will resolve to the current directory
```

The above could be prevented by using the `_isValidTaskbookCustomParentDirectory` predicate that checks both for existance and empty strings

What do you think about this?

bwl replied:
Thanks for being so thorough in your review! I tried a quick fix here decided to refactor for clarity. 

Also discovered in testing a gap in dx, where i forgot that I need to pass ~ not ~/.taskbook to pull up the main boards -- feels like an error that people may stumble into while exploring the taskboor-dir features so I added checks for that so we don't end up putting .taskbook folders inside of .taskbook folders.

Added a new error for blank --taskbook-dir flag to be a bit more helpful to users. Considered adding for the env var but I think the existing errors are covering it ok as you can't set an empty env var.

Finally, made a script to test all of the different scenarios. Just as a gist as we don't have a test or scripts dir yet. https://gist.github.com/bwl/d602e23bb6b694c1be0b696b12b31436

klaudiosinani commented on readme.md:
````diff
@@ -360,6 +363,20 @@ To search for one of more items, use the `--find`/`-f` option, followed by your
+```
+
+```
+$ TASKBOOK_DIR=~/hometasks tb
````
The new section looks really good - well done

Is the whitespace character between ~/hometasks and tb intentional?

bwl replied:
This line is showing how you can set an env var and launch tb.
I could add a comment or change the example to just show setting an env var?

bwl committed:
fix: _isValidTaskbookCustomParentDirectory method

setting --taskbook-dir without providing a value was resulting in the taskbook-dir being set to . as fs.existsSync() parses empty strings as `cwd`. This commit fixes the validation logic to prevent this issue.

bwl committed:
fix: --taskbook-dir resolution and clarify empty-flag error

Refactored the directory resolver to distinguish between parent paths
and explicit .taskbook directories, so passing a path like ~/.taskbook
now reuses the existing data instead of nesting another folder. Added
guardrails for empty --taskbook-dir flags and missing custom directories
so the CLI fails fast with clear guidance.

Test script:
https://gist.github.com/bwl/d602e23bb6b694c1be0b696b12b31436

bwl committed:
fix: sonarqube warning

klaudiosinani commented on src/directory.js:
```diff
@@ -0,0 +1,139 @@
+
+    const { taskbookDir } = options;
+
+    if (!this._isStringType(taskbookDir) || this._isEmptyString(taskbookDir)) {
```
We could use the _isPresentString predicate for this condition inverted

bwl replied:
good catch, fixed by using **existing** helper 😅

bwl committed:
refactor: add _isPresentString helper for cleaner validation

Add _isPresentString helper method and refactor validation logic
on line 73 per maintainer feedback.

Before: \!this._isStringType(x) || this._isEmptyString(x)
After: \!this._isPresentString(x)

More semantic and maintains single source of truth for
'present string' validation logic.

bwl committed:
fix: remove duplicate _isPresentString helper

Remove duplicate _isPresentString method added in previous commit.
The helper already existed at line 81, so we just need to use
the existing implementation.

klaudiosinani reviewed (approved):
@bwl Well done for the implementation and thank you for your time and effort to contribute to taskbook, your implementation will be shipped with the upcoming release : )
